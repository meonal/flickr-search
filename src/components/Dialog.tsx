import * as React from 'react';
import { Button, Modal } from 'react-bootstrap';

interface Props {
  title: string;
  message?: string;
  body?: JSX.Element;
  isShow: boolean;
  onClose: () => void;
}

class Dialog extends React.Component<Props, object> {
  render() {
    const { title, message, body, isShow, onClose } = this.props;
    return (
      <Modal show={isShow} onHide={onClose}>
        <Modal.Header closeButton>
          <Modal.Title>{title}</Modal.Title>
        </Modal.Header>
        <Modal.Body>
          {message}
          {body}
        </Modal.Body>
        <Modal.Footer>
          <Button onClick={onClose}>Close</Button>
        </Modal.Footer>
      </Modal>
    );
  }
}

export default Dialog;