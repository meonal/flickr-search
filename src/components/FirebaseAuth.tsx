import * as React from 'react';
import { authUI, uiConfig } from '../firebase/authUI';
// import 'firebaseui/dist/firebaseui.css';

const PLACEHOLDER_ID = 'firebase-auth';

interface Props {
  placeholderId?: string;
}

export default class FirebaseAuth extends React.Component<Props, any> {
  constructor(props: Props) {
    super(props);
    const { placeholderId } = props;
    this.placeholderId = placeholderId || PLACEHOLDER_ID;
  }
  placeholderId: string;

  componentDidMount() {
    authUI.reset();
    authUI.start('#' + this.placeholderId, uiConfig);
  }
  componentWillUnmount() {
    authUI.reset();
  }
  render() {
    return (
      <div id={this.placeholderId} />
    );
  }
}