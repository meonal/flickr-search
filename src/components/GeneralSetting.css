.general-setting {
  margin-left: 30px;
  width: 800px;
}

.bold {
  font-weight: bold;
}

.r {
  margin-right: 15px;
}

.margin {
  margin: 10px !important;
}