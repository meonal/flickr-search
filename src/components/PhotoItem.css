.photo-item {
  width: 200px;
  text-align: center;
  margin: 5px;
}

.photo-area {
  display: flex;
  justify-content: space-around;
  align-items: center;
  height: 200px;
  margin: 5px;
}

.photo {
  max-width: 200px;
  max-height: 200px;
}

.title {
  max-width: 200px;
  max-height: 3em;
  overflow: hidden;
  font-weight: bold;
}

.author {
  max-width: 200px;
  max-height: 3em;
  overflow: hidden;
  margin-bottom: 8px;
}

.count {
  border-radius: 3px;
  display: inline-block !important;
  margin: 0 5px;
  padding: 0 6px !important;
  font: bold 10px/18px "lucida grande", sans-serif !important;
  text-decoration: none;
  color: #0069b1 !important;
  background-color: #cceeff;
}

.space {
  margin-right: 5px;
}