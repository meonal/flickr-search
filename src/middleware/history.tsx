import { routerMiddleware } from 'react-router-redux';
import createHistory from 'history/createBrowserHistory';

const history = createHistory();

export const historyMiddleware = routerMiddleware(history);

export default history;