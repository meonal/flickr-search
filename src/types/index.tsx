import * as constants from '../constants';
import * as moment from 'moment';
import { RouterState } from 'react-router-redux';
import { UserInfo } from 'firebase';

// state

export interface State {
  search: SearchState;
  fav: FavState;
  detail: DetailState;
  setting: SettingState;
  router: RouterState;    // react-router-redux
}

export interface SearchState {
  viewType: ViewType;
  photos: PhotoItem[];
  condition: SearchCondition;
  isLoading: boolean;
}

export class SearchCondition {
  searchType: SearchType;
  text: string;
  user_id: string;
  ownername: string;
  sort: string;
  page: number;
  pages: number;
  perpage: number;
  total: string;
  message: string;

  constructor(type: SearchType) {
    this.searchType = type;
    this.text = '';
    this.user_id = '';
    this.ownername = '';
    this.sort = type === SearchType.User
      ? constants.sortPopular[0]
      : constants.sortSearch[0];
    this.page = 0;
    this.pages = 0;
    this.perpage = 25;
    this.total = '';
    this.message = '';
  }
}

export type PhotoItem = Photo & PhotoMyAppDefine;

export interface PhotoMyAppDefine {
  isFav: boolean;
}

export enum ViewType {
  Normal,
  Large,
}

export enum SearchType {
  Global,
  User,
}

export enum DetailType {
  Search = 'Search',
  Fav = 'Fav',
  None = 'None'
}
export namespace DetailType {
  export function Parse(input: string) {
    switch (input) {
      case '/':
        return DetailType.Search;
      case '/fav':
        return DetailType.Fav;
      default:
        return DetailType.None;
    }
  }
}

export interface FavState {
  viewType: ViewType;
  photos: PhotoItem[];
  // sort: string;
}

export interface DetailState {
  id: string;
  type: DetailType;
}

export enum ColorTheme {
  Light = 'light',
  Dark = 'dark',
}

export interface SettingState {
  theme: ColorTheme;
  user: UserInfo;
}

// props (View)

export class PhotoViewItem implements PhotoViewItem {
  id: string;
  title: string;
  user_id: string;
  ownername: string;
  datetaken: string;
  dateupload: string;
  lastupdate: string;
  views: number;
  url_s: string;
  url_l: string;
  url_o: string;
  pathalias: string;
  isFav: boolean;

  constructor(p: PhotoItem) {
    this.id = p.id;
    this.title = p.title;
    this.user_id = p.owner;
    this.ownername = p.ownername;
    this.datetaken = p.datetaken;
    this.dateupload = moment.unix(Number(p.dateupload)).format(constants.dateFormat);
    this.lastupdate = moment.unix(Number(p.lastupdate)).format(constants.dateFormat);
    this.views = Number(p.views);
    this.url_s = p.url_s;
    this.url_l = p.url_l;
    this.url_o = p.url_o;
    this.pathalias = p.pathalias;
    this.isFav = p.isFav;
  }
}

// Error

export class MyError extends Error {
  constructor(m: string) {
    super(m);
  }
  isRequiredRethrow: boolean;
}

// flickr api

export interface Photo {
  id: string;
  owner: string;
  secret: string;
  server: string;
  farm: number;
  title: string;
  ispublic: number;
  isfriend: number;
  isfamily: number;
  dateupload: string;
  lastupdate: string;
  datetaken: string;
  datetakengranularity: any;
  datetakenunknown: string;
  ownername: string;
  views: string;
  url_s: string;
  height_s: string;
  width_s: string;
  url_l: string;
  height_l: string;
  width_l: string;
  url_o: string;
  height_o: string;
  width_o: string;
  pathalias: string;
}

export interface Photos {
  page: number;
  pages: number;
  perpage: number;
  total: string;
  photo: Photo[];
}

export interface PhotoResult {
  photos: Photos;
  stat: string;
  code: number;     // stat === "fail"
  message: string;  // stat === "fail"
}