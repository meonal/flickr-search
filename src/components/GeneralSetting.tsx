import * as React from 'react';
import './GeneralSetting.css';
import SearchActions from '../actions/Search';
import SettingActions from '../actions/Setting';
import { ColorTheme, SettingState } from '../types';
import {
  Button, Modal, ListGroup, ListGroupItem,
  ToggleButton, ToggleButtonGroup
} from 'react-bootstrap';

interface Props {
  setting: SettingState;
  actions: {
    search: SearchActions;
    setting: SettingActions;
  };
}

class GeneralSetting extends React.Component<Props, object> {
  constructor(props: any) {
    super(props);
    this.state = {
      showModal: false,
      title: '',
      message: '',
    };
    this.close = this.close.bind(this);
    this.setTheme = this.setTheme.bind(this);
  }
  close() {
    this.setState({ showModal: false });
  }
  open() {
    this.setState({ showModal: true });
  }
  resetFav(title: string, message: string) {
    const { actions } = this.props;
    actions.search.clearFav();
    this.setState({ title, message });
    this.open();
  }
  setTheme(val: any) {
    const { actions } = this.props;
    actions.setting.setTheme(val);
  }

  render() {
    const { setting } = this.props;
    const state = this.state as any;
    return (
      <div styleName="general-setting">
        <ListGroup>
          <ListGroupItem bsStyle="info" header="テーマの切り替え">
            <div styleName="item">
              <span>アプリ全体のカラーテーマを切り替えます。</span>
              <ToggleButtonGroup type="radio" name="theme"
                defaultValue={setting.theme} onChange={this.setTheme}>
                <ToggleButton value={ColorTheme.Light}>{ColorTheme.Light}</ToggleButton>
                <ToggleButton value={ColorTheme.Dark}>{ColorTheme.Dark}</ToggleButton>
              </ToggleButtonGroup>
            </div>
          </ListGroupItem>
          <ListGroupItem bsStyle="danger" header="Favのリセット">
            <div styleName="item">
              <span>すべてのFavを削除します。この操作は戻せません。</span>
              <Button bsStyle="danger" onClick={() => this.resetFav('Setting', 'Favをリセットしました。')} >Reset</Button>
            </div>
          </ListGroupItem>
        </ListGroup>

        <Modal show={state.showModal} onHide={this.close}>
          <Modal.Header closeButton>
            <Modal.Title>{state.title}</Modal.Title>
          </Modal.Header>
          <Modal.Body>
            <h4>{state.message}</h4>
          </Modal.Body>
          <Modal.Footer>
            <Button onClick={this.close}>Close</Button>
          </Modal.Footer>
        </Modal>
      </div>
    );
  }
}

export default GeneralSetting;