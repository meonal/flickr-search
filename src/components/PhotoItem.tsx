import * as React from 'react';
import { Button, Glyphicon } from 'react-bootstrap';
import { PhotoViewItem, SearchCondition, SearchType } from '../types';
import SearchActions from '../actions/Search';
import { withRouter } from 'react-router-dom';
import './PhotoItem.css';

interface Props {
  item: PhotoViewItem;
  actions: SearchActions;
  index: number;
  onPhotoClick: (index: number) => void;
}

class PhotoItem extends React.Component<Props, object> {
  condition = new SearchCondition(SearchType.User);

  render() {
    const { item, actions, index, onPhotoClick } = this.props;
    const pathname = (this.props as any).location.pathname;

    const bsStyle = item.isFav ? 'success' : 'primary';
    const star = 'star' + (item.isFav ? '' : '-empty');

    return (
      <div styleName="photo-item">
        <div styleName="photo-area">
          <a href="#" onClick={(e: any) => { onPhotoClick(index); e.preventDefault(); }}>
            <img styleName="photo" src={item.url_s} alt={item.title} />
          </a>
        </div>
        <div styleName="title">{item.title}</div>
        <div styleName="author">
          <a href="#" onClick={() => actions.fetchPhotoTransition(
            { ...this.condition, user_id: item.user_id, ownername: item.ownername }, pathname)}>
            {item.ownername}</a>
        </div>
        <div styleName="count">{item.views} Views</div>
        <Button styleName="space" bsSize="small" onClick={() => actions.gotoDetail(item.id, pathname)} >
          <Glyphicon glyph="info-sign" />
        </Button>
        <Button bsStyle={bsStyle} bsSize="small" onClick={() => actions.toggleFav(item.id)} >
          <Glyphicon glyph={star} />
        </Button>
      </div>
    );
  }
}

export default withRouter(PhotoItem);