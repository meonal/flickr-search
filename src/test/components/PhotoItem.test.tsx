import * as React from 'react';
import * as renderer from 'react-test-renderer';
import * as enzyme from 'enzyme';
import * as PhotoItem from '../../components/PhotoItem'
import { PhotoViewItem } from '../../types/index';
import SearchActions from '../../actions/Search';
import configureStore from '../../store/configureStore';
import { Location } from 'history';

export const store = configureStore();

const item: PhotoViewItem = {
  id: "1",
  title: "photo title",
  user_id: "2",
  ownername: "foo bar",
  datetaken: "2017/01/01 12:34:56",
  dateupload: "2017/02/01 12:34:56",
  lastupdate: "2017/03/01 12:34:56",
  views: 100,
  url_s: "http://example.com/s",
  url_l: "http://example.com/l",
  url_o: "http://example.com/9",
  pathalias: "hoge/fuga",
  isFav: true,
};

const actions = SearchActions.getInstance(store.dispatch);
const onPhotoClick = (index: number) => { };
const location: Location = {
  pathname: '/fav',
  search: '',
  state: '',
  hash: '',
  key: ''
}

describe('スナップショット', () => {
  it('スナップショット', () => {
    const snapshot = renderer.create(
      <PhotoItem.PhotoItem item={item} actions={actions} index={1}
        onPhotoClick={onPhotoClick} location={location} />)
      .toJSON();
    expect(snapshot).toMatchSnapshot();
  });
});

describe('テキストレンダリング', () => {
  let photoItem: enzyme.ShallowWrapper<any, any>;
  beforeEach(() => {
    photoItem = enzyme.shallow(
      <PhotoItem.PhotoItem item={item} actions={actions} index={1}
        onPhotoClick={onPhotoClick} location={location} />);
  });

  it('title', () => {
    expect(photoItem.find('[styleName="title"]').text()).toEqual('photo title');
    expect(photoItem.find('[styleName="title"]').text()).not.toEqual('');
  });
  it('author', () => {
    expect(photoItem.find('[styleName="author"]').text()).toEqual('foo bar');
    expect(photoItem.find('[styleName="author"]').text()).not.toEqual('');
  });
  it('count', () => {
    expect(photoItem.find('[styleName="count"]').text()).toEqual('100 Views');
    expect(photoItem.find('[styleName="count"]').text()).not.toEqual('0 Views');
    expect(photoItem.find('[styleName="count"]').text()).not.toEqual('');
  });
});