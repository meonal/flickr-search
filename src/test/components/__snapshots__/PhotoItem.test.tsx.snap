// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`スナップショット 1`] = `
<div
  styleName="photo-item"
>
  <div
    styleName="photo-area"
  >
    <a
      href="#"
      onClick={[Function]}
    >
      <img
        alt="photo title"
        src="http://example.com/s"
        styleName="photo"
      />
    </a>
  </div>
  <div
    styleName="title"
  >
    photo title
  </div>
  <div
    styleName="author"
  >
    <a
      href="#"
      onClick={[Function]}
    >
      foo bar
    </a>
  </div>
  <div
    styleName="count"
  >
    100
     Views
  </div>
  <button
    className="btn btn-sm btn-default"
    disabled={false}
    onClick={[Function]}
    styleName="space"
    type="button"
  >
    <span
      className="glyphicon glyphicon-info-sign"
    />
  </button>
  <button
    className="btn btn-sm btn-success"
    disabled={false}
    onClick={[Function]}
    type="button"
  >
    <span
      className="glyphicon glyphicon-star"
    />
  </button>
</div>
`;

exports[`スナップショット スナップショット 1`] = `
<div
  styleName="photo-item"
>
  <div
    styleName="photo-area"
  >
    <a
      href="#"
      onClick={[Function]}
    >
      <img
        alt="photo title"
        src="http://example.com/s"
        styleName="photo"
      />
    </a>
  </div>
  <div
    styleName="title"
  >
    photo title
  </div>
  <div
    styleName="author"
  >
    <a
      href="#"
      onClick={[Function]}
    >
      foo bar
    </a>
  </div>
  <div
    styleName="count"
  >
    100
     Views
  </div>
  <button
    className="btn btn-sm btn-default"
    disabled={false}
    onClick={[Function]}
    styleName="space"
    type="button"
  >
    <span
      className="glyphicon glyphicon-info-sign"
    />
  </button>
  <button
    className="btn btn-sm btn-success"
    disabled={false}
    onClick={[Function]}
    type="button"
  >
    <span
      className="glyphicon glyphicon-star"
    />
  </button>
</div>
`;
